### PR TITLE
8340824: C2: Memory for TypeInterfaces not reclaimed by hashcons()

### PR DESCRIPTION
**Pull request by rwestrel**
The list of interfaces for a `TypeInterfaces` is contained in a `GrowableArray` that's allocated in the type arena. When `hashcons()` deletes a `TypeInterfaces` object because an identical one exists, it can't reclaim memory for the object because it can only free the last thing that was allocated and that's the backing store for the `GrowableArray`, not the `TypeInterfaces` object.

This patch changes the array of interfaces stored in `TypeInterfaces` into a pointer to a `GrowableArray`. `TypeInterfaces::make` calls `hashcons` with a temporary copy of the array of interfaces allocated in the current thread's resource area. This way if `hascons` deletes the `TypeInterfaces`, it is the last thing allocated in the type arena and memory can be reclaimed.  Memory for the `GrowableArray` is freed as well on return from `TypeInterfaces::make`. If the newly allocated `TypeInterfaces` survives `hashcons` then a permanent array of interfaces is allocated in the type arena and linked from the `TypeInterfaces` object.

I ran into this issue while working on a fix for a separate issue that causes more Type objects to be created. With that prototype fix, TestScalarReplacementMaxLiveNodes uses 1.4GB of memory at peak. With the patch I propose here on top, memory usage goes down to 150-200 MB which is in line with the peak memory usage for TestScalarReplacementMaxLiveNodes when run with current master.

When I opened the PR initially, the fix I proposed was to let `GrowableArray` try to reclaim memory from an arena when destroyed. With that patch, some gtests failed when a `GrowableArray` is created by a copy constructor and the backing store for the array is shared. As a consequence, I reconsidered the fix and thought it was safer to go with a fix that only affects `TypeInterfaces`.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340824](https://bugs.openjdk.org/browse/JDK-8340824): C2: Memory for TypeInterfaces not reclaimed by hashcons() (**Bug** - P3)

### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21163/head:pull/21163` \
`$ git checkout pull/21163`

Update a local copy of the PR: \
`$ git checkout pull/21163` \
`$ git pull https://git.openjdk.org/jdk.git pull/21163/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21163`

View PR using the GUI difftool: \
`$ git pr show -t 21163`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21163.diff">https://git.openjdk.org/jdk/pull/21163.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21163#issuecomment-2371705701)

### CODE REVIEW

**Comment by rwestrel** [comment]
Converted back to draft while I investigate gtest failures.

File: src/hotspot/share/opto/type.cpp [hunk 1]
```diff
@@ -3260,24 +3260,38 @@ void TypeRawPtr::dump2( Dict &d, uint depth, outputStream *st ) const {
...
-    add(interfaces->at(i));
-  }
+  assert(_interfaces != nullptr, "no interfaces should be passed as empty list");
   initialize();
 }
 
-const TypeInterfaces* TypeInterfaces::make(GrowableArray<ciInstanceKlass*>* interfaces) {
-  TypeInterfaces* result = (interfaces == nullptr) ? new TypeInterfaces() : new TypeInterfaces(interfaces);
-  return (const TypeInterfaces*)result->hashcons();
+const TypeInterfaces* TypeInterfaces::make(const GrowableArray<ciInstanceKlass*>* interfaces) {
```
Comment (merykitty) [inline src/hotspot/share/opto/type.cpp:3270]: I think you can make `_interface` a `ciInstanceKlass**` and do this:

    void* ptr = Type::operator new(sizeof(TypeInterfaces) + sizeof(ciInstanceKlass*) * interfaces->length())

Then `delete ptr` should drop the whole thing.

Comment (merykitty) [reply src/hotspot/share/opto/type.cpp:3270]: A `GrowableArrayFromArray` would be mostly compatible with the interface of `GrowableArray`, too.

Comment (rwestrel) [reply src/hotspot/share/opto/type.cpp]: Ah! nice. I wasn't aware of `GrowableArrayFromArray`. Updated change follows your suggestion.

**Review by iwanowww — APPROVED** [review]
Looks good.

It feels a bit weird to see `GrowableArray` used to represent a read-only data structure, but I understand that you still benefit from some helper methods it provides.

**Review by merykitty — APPROVED** [review]

**Comment by rwestrel** [comment]
@iwanowww @merykitty thanks for the reviews

**Pull request closed without merging**